### PR TITLE
[Feature: #162727565] Add search filter functionality

### PR DESCRIPTION
**Pull request by rajeman**
#### What does this PR do?
- It ensures articles can be searched for
#### Description of Task to be completed?
- [x] make the `GET /articles`  generic
- [x] ensure all articles are returned if no filter is provided
- [x] ensure only articles that match the specified filters will be returned
#### How should this be manually tested?
- goto `GET /articles` to view all articles and their details
- goto `GET /articles?filter=value` to view article with specified filter, where filter can be `author`, `tag`, `keywords`, or `category`
- you can supply as many filters as available
- route is not protected
#### What are the relevant pivotal tracker stories?
- [#162727565](https://www.pivotaltracker.com/n/projects/2229853/stories/162727565)

### CODE REVIEW

File: controllers/article.js [hunk 1]
```diff
@@ -86,48 +94,77 @@ export default {
...
+      }
+    ];
+    if (author) {
+      queryArray.push({
+        [or]: {
+          '$User.Profile.lastName$': {
+            [iLike]: `%${author}%`
+          },
+          '$User.Profile.firstName$': {
+            [iLike]: `%${author}%`
```
Comment (oreoluwade) [inline controllers/article.js:128]: We should probably add `username` here too.

File: controllers/article.js [hunk 2]
```diff
@@ -86,48 +94,77 @@ export default {
...
+          },
+        }
+
+      });
+    }
     const article = await Article.findAll({
```
Comment (oreoluwade) [inline controllers/article.js]: Help change this to `articles` to appropriately connote multiple results expected.

File: test/011-article-search.test.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+import chaiHttp from 'chai-http';
+import models from '../models';
+import server from '../app';
+
+const baseUrl = '/api/v1';
+const articleUrl = '/api/v1/articles';
+
+chai.use(chaiHttp);
+
+describe('ARTICLE TEST SUITE', () => {
```
Comment (oreoluwade) [inline test/011-article-search.test.js]: We can change this to `GET MULTIPLE ARTICLES`

File: test/011-article-search.test.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      });
+
+    it('should get all articles with supplied tag and author',
+      async () => {
+        const response = await chai.request(server)
+          .get(`${articleUrl}?tag=skills&author=jane`);
+        expect(response.body.count).to.equal(1);
+      });
+
+    it('should get all articles with all filters applied',
```
Comment (oreoluwade) [inline test/011-article-search.test.js]: `should combine multiple filters in getting just the matching article(s)`

**Review by oreoluwade — CHANGES_REQUESTED** [review]
Good work. Just a couple of comments.

Also, you'd need to fix some merge conflicts in the articles controller I think.

**Review by oreoluwade — APPROVED** [review]
💯

**Comment by oreoluwade** [comment]
One last thing. Could you merge the two `/` routes fir articles? There is get, and there is post and they are currently apart.

**Comment by rajeman** [comment]
I have done that

**Pull request merged**